packages/wrangler/src/api/startDevWorker/BundlerController.ts

```diff
@@ -1,37 +1,280 @@
+				},
+			});
+		} catch (err) {
+			logger.error("Custom build failed:", err);
+			this.emitErrorEvent({
+				type: "error",
+				reason: "Custom build failed",
+				cause: castErrorCause(err),
+				source: "BundlerController",
+				data: undefined,
+			});
+		}
 	}
 
-	async teardown() {
-		notImplemented(this.teardown.name, this.constructor.name);
+	#startCustomBuild(config: StartDevWorkerOptions) {
+		assert(config._entry);
+		assert(config.build?.custom.watch);
+		this.#customBuildWatcher = watch(config.build?.custom.watch, {
+			persistent: true,
+			ignoreInitial: true,
+		});
+		this.#customBuildWatcher.on(
+			"ready",
+			() =>
+				void this.#runCustomBuild(config, String(config.build?.custom.watch))
+		);
+
+		this.#customBuildWatcher.on(
+			"all",
+			(_event, filePath) => void this.#runCustomBuild(config, filePath)
+		);
+	}
+
+	#bundlerCleanup?: ReturnType<typeof runBuild>;
+
+	#startBundle(config: StartDevWorkerOptions) {
+		assert(this.#tmpDir);
+		assert(config._entry, "config._entry");
+		assert(config._additionalModules, "config._additionalModules");
+		assert(config.build?.moduleRules, "config.build?.moduleRules");
+		assert(config.build?.define, "config.build?.define");
+		this.#bundlerCleanup = runBuild(
+			{
+				entry: config._entry,
+				destination: this.#tmpDir.path,
+				jsxFactory: config.build?.jsxFactory,
+				jsxFragment: config.build?.jsxFragment,
+				processEntrypoint: Boolean(config._processEntrypoint),
+				additionalModules: config._additionalModules,
+				rules: config.build.moduleRules,
+				assets: config._assets,
+				serveAssetsFromWorker: Boolean(config._serveAssetsFromWorker),
+				tsconfig: config.build?.tsconfig,
+				minify: config.build?.minify,
+				nodejsCompatMode: config.build.nodejsCompatMode,
+				define: config.build.define,
+				noBundle: !config.build?.bundle,
+				findAdditionalModules: config.build?.findAdditionalModules,
+				durableObjects: config._bindings?.durable_objects ?? { bindings: [] },
+				local: !config.dev?.remote,
+				// startDevWorker only applies to "dev"
+				targetConsumer: "dev",
+				testScheduled: Boolean(config.dev?.testScheduled),
+				projectRoot: config._projectRoot,
+				onStart: () => {
+					this.emitBundleStartEvent({
+						type: "bundleStart",
+						config,
+					});
+				},
+				defineNavigatorUserAgent: isNavigatorDefined(
+					config.compatibilityDate,
+					config.compatibilityFlags
+				),
+			},
+			(cb) => {
+				const newBundle = cb(this.#currentBundle);
+				this.emitBundleCompletedEvent({
+					type: "bundleComplete",
+					config,
+					bundle: newBundle,
+				});
+				this.#currentBundle = newBundle;
+			},
+			(err) =>
+				this.emitErrorEvent({
+					type: "error",
+					reason: "Failed to construct initial bundle",
+					cause: castErrorCause(err),
+					source: "BundlerController",
+					data: undefined,
+				})
+		);
+	}
+
+	#tmpDir?: EphemeralDirectory;
+
+	async onConfigUpdate(event: ConfigUpdateEvent) {
+		await this.teardown();
```

Comment:
This isn't how teardown should be used – ideally it should only be called when the dev env is not expected to be used anymore. And this also causes onConfigUpdate to become async.

Ideally, each step would manage itself more granularly. This allows only the parts that need to be async to be async and reduce unnecessary async boundaries

Reply:
That's fair—do you have a suggestion for how this could/should be done differently? The things that need done on every config update are the same as those that need done when doing a teardown, and unfortunately some of those are async (closing the esbuild and filesystem watchers)